---
title: "CAT: The Cell Authentication Toolkit"
author: "Erik Fasterius"
# abstract: "This is the vignette for the Cell Authentication Toolkit (CAT) R
          # package."
date: "`r Sys.Date()`"
output:
    BiocStyle::pdf_document:
        highlight: tango
bibliography: bibliography.bib
vignette: >
    %\VignetteIndexEntry{CAT: The Cell Authentication Toolkit}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

# Introduction

This vignette describes a workflow of using the \Rpackage{"CellAuthentication"}
package for authentication, characterisation and evaluation of two or more
high throughput sequencing (*i.e.* RNA-seq or whole genome sequencing) samples.
The general principle is to create *single nucelotide variant* (SNV) profiles
of every sample of interest, followed by comparisons between each set to find
their overall similarity, in addition to detailed analyses of the differences.
By analysing your data with this workflow you will not only be able to
authenticate your samples to a high degree of confidence, but you will also be
able to investigate what genes and transcripts are affected by SNVs differing
between your samples, what biological effect they will have, which pathways
they belong to, and more. The workflow consists of three separate steps:

    1.  Creation of SNV profiles
    2.  SNV profile comparisons
    3.  Authentication, characterisation and evaluation

Each step has its own section below demonstrating how to perform the analyses.
Input data should be in the form of [VCF files][1], *i.e* output from variant
callers such as the [Genome Analysis ToolKit][2].

[1]: http://www.internationalgenome.org/wiki/Analysis/variant-call-format
[2]: https://software.broadinstitute.org/gatk/

## Installation

The latest stable release of this package can be found on [BioConductor][3] and
installed using the `biocLite` function:

```{r Installation, eval = FALSE}
source("https://bioconductor.org/biocLite.R")
biocLite("CellAuthentication")
```

This will also install any missing packages requires for full functionality,
should they not already exist in your system. If you haven't installed
BioConductor, you can do so by simply calling `biocLite()` without specifying a
package, and it will be installed for you. You can read more about this at
BioConductor's [installation page][4].

[3]: http://bioconductor.org/
[4]: http://bioconductor.org/install/

# Creation of SNV profiles

The first step of the workflow is to create the SNV profile of each sample,
which can then be compared to each other. While computation time is usually not
an issue for simple binary comparisons (*i.e.* comparisons with only two
samples), this can quickly become a concern for analyses where samples are
compared to several others (*i.e.* A vs B, A vs C, ..., A vs Z). In order to
decrease the computation time for large comparison sets and to facilitate
re-analyses with different parameters each SNV profile is saved on disc as a
normal `.txt` file.

The creation of a SNV profile includes filtering of low-confidence variants and
removal of variants below a sequencing depth threshold (`10` by default). Only
records with the highest SNV impact (*i.e.* putative impact on protein 
function) for each variant is kept, as they are most likely to affect the
biology of the cells.

## Create profiles with R

The example data used in the demonstration can be found in the `inst/extdata/`
directory.

```{r Create SNV profile with R}
# Load the package
library("CellAuthentication")

# List the example VCF file
file <- system.file("extdata", "example.vcf.gz",
                    package = "CellAuthentication")

# Create two SNV profiles
create_profile(file, "sample1", "profile_1.txt")
create_profile(file, "sample2", "profile_2.txt", filter_depth = 15)
```

This creates SNV profiles for the two samples found in the example data
(`sample1` and `sample2`) and saves them as `profile_1.txt` and `profile_2.txt`
in the current directory, respectively. The profile of the second sample was
created with a non-standard filter for sequencing depth (`15`), which should
only be done if you want a stricter criteria for your profile.

## Faster profile creation with Python

SNV profiles can also be created with Python, another scripting langage, if you
have it installed. The `create_profile` function also doubles as a wrapper
for the `create_profile.py` script included in the package, which can create
SNV profiles much quicker than its `R` equivalent. This might not be important
for a lot of users, but is nevertheless included for cases where it helps to
have the extra speed.

```{r Create SNV profile with Python}
create_profile(file, "sample2", "profile_2.txt", python = TRUE)
```

In order to run the `Python` SNV creation script, you will need to have the
[PyVCF](https://pypi.python.org/pypi/PyVCF) installed. You can also run the
python script directly from the command line, supplying its input arguments
from there. To see the arguments, use:

```{r Python command line, engine = "bash", eval = FALSE}
$ python create_profile.py --help
```

# SNV profile comparisons

Once each relevant sample has its own SNV profile the comparisons can be
performed. First, each profile is read using the `read_profile` function,
which outputs `GRanges` objects for fast and efficient overlapping of variants.

```{r Read SNV profiles}
profile1 <- read_profile("profile_1.txt", "sample1")
profile2 <- read_profile("profile_2.txt", "sample2")
head(profile1)
```

```{r Remove temporary files, echo = FALSE, results = "hide"}
file.remove("profile_1.txt")
file.remove("profile_2.txt")
```

Once each profile has been read, the overlapping variants between them can be
found using the `overlap_profiles` function. Only variants found in both
profiles are considered to overlap, as similarity calculations between profiles
where some variants only have confident calls in one of the samples are
inappropriate.

```{r Overlap SNV profiles}
overlaps <- overlap_profiles(profile1, profile2)
```

The final step is to see which SNVs match between the two profiles being
compared, which is accomplished with the `compare_overlaps` function. A SNV is
considered a match if it has an identical genotype in both profiles. 

```{r Compare profiles}
overlaps <- compare_overlaps(overlaps)
```

We can look at the resulting data frame:

```{r Inspect comparison}
overlaps[1:5, c(1, 2, 5, 19, 20, 35, 36)]
```

You are now ready for downstream analyses!

# Authentication, characterisation and evaluation

## Similarity and global statistics

When you finally have your matched, overlapping SNVs, it's time to analyse and
characterise them. The first thing you might want to check are the global
similarities and summary statistics, which can be done with the
`calculate_similarity` function. The `concordance` is simply the number of
matching genotypes divided by the total number of overlapping variants, while
the `similarity score` is a weighted measure of the concordance in the form of
a binomial experiment, taking into account the number of overlapping variants
available:

$$Similarity = \frac{s + a}{n + a + b}$$

... where `s` is the number of matching genotypes, `n` is the total number of 
overlapping SNVs, `a` and `b` being the parameters used to weigh the
concordance in favour of comparisons with more overlaps. The default
parameters of `1` and `5` were selected to yield an equivalent cutoff to a
previously used one [@Yu2015], which results in a lower limit 44 of perfectly
matching overlapping variants with a similarity score of 90. The similarity
score is thus a better measure of biological equivalency than just the
concordance.

```{r Calculate similarities}
similarity <- calculate_similarity(overlaps)
similarity
```

Here, you can see a summary of the relevant statistics for your particular
comparison: the number of overlaps between your two samples, the number of
matching genotypes, their concordance as well as their similarity score. You
may additionally change the parameters of the score (if you, for example, want
a stricter calculation). You may also supply the `calculate_similarity`
function with an existing dataframe with summary data produced previously, in
order to aggregate scores and statistics for an arbitrary number of
comparisons:

```{r Calculate similarities iteratively}
similarities <- calculate_similarity(overlaps, similarity, a = 1, b = 10)
similarities
```

Notice that the new `similarities` dataframe contains two rows, but the
second one has a lower score due to the changing of the `b` parameter from `1`
to `10`. Any number of samples can be added using this function, for use in
further downstream  analyses.

# Performance

# Citation

If you are using CAT to analyse your samples, please cite it formally:

Fasterius, E., Raso, C., Kennedy, S., Rauch, N., Lundin, P., Kolch, W., et al.
(2017). A novel RNA sequencing data analysis method for cell line
authentication. PloS One, 12(2), e0171435.
http://doi.org/10.1371/journal.pone.0171435

# Session info
```{r Session info, echo = FALSE}
sessionInfo()
```

# References
