---
title: "seqCAT: The High Throughput Sequencing Cell Authentication Toolkit"
author: "Erik Fasterius"
date: "`r Sys.Date()`"
output:
    BiocStyle::html_document:
        highlight: tango
        toc: true
        number_sections: true
bibliography: bibliography.bib
vignette: >
    %\VignetteIndexEntry{seqCAT: The High Throughput Sequencing Cell Authentication Toolkit}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

```{r Options, echo = FALSE}
knitr::opts_chunk$set(fig.align = "center")
```

# Introduction

This vignette describes a workflow of using the \Rpackage{"seqCAT"}
package for authentication, characterisation and evaluation of two or more
high throughput sequencing (*i.e.* RNA-seq or whole genome sequencing) samples.
The general principle is to create *single nucelotide variant* (SNV) profiles
of every sample of interest, followed by comparisons between each set to find
their overall similarity, in addition to detailed analyses of the differences.
By analysing your data with this workflow you will not only be able to
authenticate your samples to a high degree of confidence, but you will also be
able to investigate what genes and transcripts are affected by SNVs differing
between your samples, what biological effect they will have, and more. The
workflow consists of three separate steps:

    1.  Creation of SNV profiles
    2.  Comparisons of SNV profiles
    3.  Authentication, characterisation and evaluation of comparisons

Each step has its own section(s) below demonstrating how to perform the
analyses. Input data should be in the form of [VCF files][1], *i.e* output from
variant callers such as the [Genome Analysis ToolKit][2] and annotated with
software such as [SnpEff][3].

[1]: http://www.internationalgenome.org/wiki/Analysis/variant-call-format
[2]: https://software.broadinstitute.org/gatk/
[3]: http://snpeff.sourceforge.net/

## Installation

The latest stable release of this package can be found on [BioConductor][3] and
installed using the `biocLite` function:

```{r Installation, eval = FALSE}
source("https://bioconductor.org/biocLite.R")
biocLite("seqCAT")
```

This will also install any missing packages requires for full functionality,
should they not already exist in your system. If you haven't installed
BioConductor, you can do so by simply calling `biocLite()` without specifying a
package, and it will be installed for you. You can read more about this at
BioConductor's [installation page][4].

[4]: http://bioconductor.org/
[5]: http://bioconductor.org/install/

You can find the development version of this package on [GitHub][6]. Please
note that it may contain bugs and work-in-progress.

[6]: https://github.com/fasterius/seqCAT

# Creation of SNV profiles

The first step of the workflow is to create the SNV profile of each sample,
which can then be compared to each other. While computation time is usually not
an issue for simple binary comparisons (*i.e.* comparisons with only two
samples), this can quickly become a concern for analyses where samples are
compared to several others (A vs B, A vs C, ..., and so on). In order to
decrease the computation time for large comparison sets and to facilitate
re-analyses with different parameters each SNV profile is saved on disc as a
normal `.txt` file.

The creation of a SNV profile includes filtering of low-confidence variants and
removal of variants below a sequencing depth threshold (`10` by default). Only
records with the highest SNV impact (*i.e.* putative impact on protein 
function) for each variant is kept, as they are most likely to affect the
biology of the cells.

## Create profiles with R

Throughout this vignette we will be using some example data, `example.vcf.gz`,
which comes from the initial publication of the general process of this method.
[@Fasterius2017] It is a simplified multi-sample VCF file on a subset of
chromosome 12 (containing all variants up to position `25400000`, in order to
keep the file size low) for three different colorectal cancer cell lines:
*HCT116*, *HKE3* and *RKO*.

```{r Create SNV profile with R}
# Load the package
library("seqCAT")

# List the example VCF file
vcf <- system.file("extdata", "example.vcf.gz",
                   package = "seqCAT")

# Create two SNV profiles
create_profile(vcf, "HCT116", "hct116_profile.txt")
create_profile(vcf, "RKO", "rko_profile.txt", filter_depth = 15)
```

This creates SNV profiles for the two samples found in the example data
(`HCT116` and `RKO`) and saves them as `hct116.profile` and `rko_profile.txt`
in the current directory, respectively. The profile of the second sample was
created with a non-standard filter for sequencing depth (`15`), which should
only be done if you want a stricter criteria for your profile.

## Faster profile creation with Python

SNV profiles can also be created with Python, another scripting langage, if you
have it installed. The `create_profile` function also doubles as a wrapper
for the `create_profile.py` script included in the package, which can create
SNV profiles much quicker than its `R` equivalent. This might not be important
for a lot of users, but is nevertheless included for cases where it helps to
have the extra speed.

```{r Create SNV profile with Python, eval = FALSE}
create_profile(vcf, "RKO", "RKO_profile.txt", python = TRUE)
```

In order to run the `Python` SNV profile creation script, you will need to have
the [PyVCF](https://pypi.python.org/pypi/PyVCF) module installed. You can also
run the python script directly from the command line, supplying its input
arguments from there. To see the arguments, use:

```{r Python command line, engine = "bash", eval = FALSE}
$ python create_profile.py --help
```

## Create COSMIC profiles

While looking at the entirety of either the genome (whole genome sequencing) or
transcriptome (RNA sequencing) and using the variants found therein gives
unprecedented statistical power and great opportunities for functional
evaluations, there are also good reasons to compare your samples' variants to
some external source. Such a source is the *Catalogue of somatic mutations in
cancer*, or *COSMIC*. COSMIC has over a thousand cell line-specific mutational
profiles, and is thus a very useful resource if you are working with cell
lines.

In order to use the COSMIC cell line database, you need to sign up for an
account at their [website][6] and get permission to download their files (which
is given free of charge to academia and non-profit organisation, but requires a
commersial license for for-profit organisations). The file you are looking for
is the `CosmicCLP_MutantExport.tsv.gz`, listed under *complete mutational data*
[here][7]. As redistributing this file is not allowed, this package includes an
extremely minimal subset of the original file, only useful for examples in this
vignette and unit testing. *Do not* use this file for your own analyses, as
your results will neither be complete nor correct!

[7]: http://cancer.sanger.ac.uk/cosmic
[8]: http://cancer.sanger.ac.uk/cell_lines/download

The first thing you might want to do is to see if your specific cell line is
available in the database, which can be accomplished using the `list_cosmic`
function:

```{r List COSMIC}
file <- system.file("extdata", "subset_CosmicCLP_MutantExport.tsv.gz",
                    package = "seqCAT")
cell_lines <- list_cosmic(file)
head(cell_lines)
```

This gives us a simple vector containing all the available cell lines in the
COSMIC database (this version of the file is for the GRCh37 assembly). You can
search it for a cell line of your choice:

```{r Search COSMIC}
any(grepl("HCT116", cell_lines))
```

All COSMIC-related functions perform some simplification of cell line names (as
there is variation in the usage of dashes, dots and other symbols), and are
case-insensitive. When you have asserted that your cell line of interest is
available, you can then read the profile for that cell line using the
`read_cosmic` function:

```{r Read COSMIC}
cosmic <- read_cosmic(file, "HCT116")
head(cosmic)
```

You now have a small, COSMIC SNV profile for your cell line, which you can
compare to any other profile you may have data for (more on this below). You
can also check how many variants are listed in COSMIC for your particular cell:

```{r Count COSMIC}
length(cosmic)
```

Here we only see a single variant for the HCT116 cell line, which is only
because of the extreme small subset of the actual file being used here. HCT116
has, in fact, over 2000 listed COSMIC SNVs, making it one of the more
abundantly characterised cell lines available (as most cell lines has only a
few hundred SNVs listed in COSMIC).

# Comparing SNV profiles

## Comparing full profiles

Once each relevant sample has its own SNV profile the comparisons can be
performed. First, each profile is read using the `read_profile` function,
which outputs `GRanges` objects for fast and efficient comparisons.

```{r Read SNV profiles}
hct116 <- read_profile("hct116_profile.txt", "HCT116")
rko <- read_profile("rko_profile.txt", "RKO")
head(hct116)
```

SNV profiles contain most of the relevant annotation data from the original VCF
file, including SNV impacts, gene/transcript IDs, mutational (rs) ID, and so
on. The `DP` (depth) field lists the total sequencing depth of this variant,
while the specific allelic depths can be found in `AD1` and `AD2`. The alleles
of each variant can be found in `A1` and `A2`.

Once each profile has been read, the genotypes of the overlapping variants
between them can be compared using the `compare_profiles` function. Only
variants found in both profiles are considered to overlap, as similarity
calculations between profiles where some variants only have confident calls in
one of the samples are inappropriate. An SNV is considered a match if it has an
identical genotype in both profiles.

```{r Compare profiles}
hct116_rko <- compare_profiles(hct116, rko)
head(hct116_rko)
```

The resulting dataframe retains all the information from each input profile
(including any differing annotation, should they exist), and lists the depths
and alleles by adding the sample names as suffixes to the relevant column
names.

## Comparing to COSMIC profiles

If you are working with cell lines and only want to analyse a subset of your
data or as a orthogonal method complementary to others, you could compare your
profile to a COSMIC profile. This works much in the same way as comparing to
another full profile, but gives slightly different output:

```{r Compare with COSMIC}
hct116_cosmic <- compare_profiles(hct116, cosmic)
head(hct116_cosmic)
```

You can use all the functions for downstream analyses for comparisons with
COSMIC data as you would full SNV profiles, but your options for functional
analyses will be limited, given that the COSMIC database is biased towards
well-known and characterised mutations. It is, however, an excellent way to
authenticate your cell lines and to assert the status of those mutations that
*should* exist in your cells.

# Evaluating binary comparisons

## Similarity and global statistics

When you finally have your matched, overlapping SNVs, it's time to analyse and
characterise them. The first thing you might want to check are the global
similarities and summary statistics, which can be done with the
`calculate_similarity` function. The `concordance` is simply the number of
matching genotypes divided by the total number of overlapping variants, while
the `similarity score` is a weighted measure of the concordance in the form of
a binomial experiment, taking into account the number of overlapping variants
available:

$$Similarity = \frac{s + a}{n + a + b}$$

... where `s` is the number of matching genotypes, `n` is the total number of 
overlapping SNVs, `a` and `b` being the parameters used to weigh the
concordance in favour of comparisons with more overlaps. The default
parameters of `1` and `5` were selected to yield an equivalent cutoff to a
previously used one [@Yu2015], which results in a lower limit 44 of perfectly
matching overlapping variants with a similarity score of 90. The similarity
score is thus a better measure of biological equivalency than just the
concordance.

```{r Calculate similarities}
similarity <- calculate_similarity(hct116_rko)
similarity
```

Here, you can see a summary of the relevant statistics for your particular
comparison: the number of overlaps between your two samples, the number of
matching genotypes, their concordance as well as their similarity score. The
cutoff used by Yu *et al.* for cell line authenticity was `90 %` for their 48
SNP panel, something that would be considered the baseline for this method as
well. The score, `68.7`, is well below that cutoff, and we can thus be certain
that these two cells are indeed not the same (as expected). While a score just
below `90` does not mean that the cell definitely are different, it *does* mean
that more rigorous evaluation needs to be performed in order to ensure their
biological equivalency.

You may additionally change the parameters of the score (if you, for example,
want a stricter calculation). You may also supply the `calculate_similarity`
function with an existing dataframe with summary data produced previously, in
order to aggregate scores and statistics for an arbitrary number of
comparisons.

```{r Calculate similarities iteratively}
# Create and read HKE3 profile
create_profile(vcf, "HKE3", "hke3_profile.txt")
hke3 <- read_profile("hke3_profile.txt", "HKE3")

# Compare HCT116 and HKE3
hct116_hke3 <- compare_profiles(hct116, hke3)

# Add HCT116/HKE3 similarities to HCT116/RKO similarities
similarities <- calculate_similarity(hct116_hke3,
                                     similarity, a = 1, b = 10)
similarities
```

Notice that the new `similarities` dataframe contains both the comparisons of
HCT116/RKO and HCT116/HKE3, and we can clearly see that HCT116 and HKE3 are
indeed very similar, as expected (HKE3 was derived from HCT116). This is true
even when using a higher value for the `b` parameter. Any number of samples can
be added using the `calculate_similarity` function, for use in further
downstream analyses.

## Evaluation of SNV impacts

An SNV's *impact* represent the putative effect that variant may have on the
function of the resulting protein, and ranges from HIGH through MODERATE, LOW
and MODIFIER, in decreasing order of magnitude. HIGH impact variants may, for
example, lead to truncated proteins due to the introduction of a stop codon,
while MODIFIER variants have little to no effect on the protein at all. It is
informative to look at the distribution of the impacts of the overlapping SNVs
between two profiles, a task easily performed by the `plot_impacts` function:

```{r Impact distributions}
impacts <- plot_impacts(hct116_rko)
impacts
```

This function takes a comparison dataframe as input and simply plots the impact
distribution of the overlapping variants. It has a number of arguments with
defaults, such as if you want to add text with the actual numbers to the plot
(`annotate = TRUE` by default), if you want to show the legend (`legend =
TRUE` by default) and what colours you want to plot the match-categories with
(`palette = c("#0D2D59", "#1954A6")` by default, two nice shades of blue). We
can see that most of the SNVs are present in the MODIFIER impact category, and
that there is not a single mismatched HIGH impact SNV. (You can also visualise
the impact distribution between your sample and the COSMIC database in exactly
the same way.)

You might also want to look at only a subset of variants, *e.g.* only the
variants with HIGH or MODERATE impacts, which is easily achieved with some data
manipulation:

```{r Subset impacts}
hct116_rko_hm <- hct116_rko[hct116_rko$impact == "HIGH" |
                            hct116_rko$impact == "MODERATE", ]
nrow(hct116_rko_hm)
```

## Evaluation of specific chromosomes, regions, genes and transcripts

You might be interested in a specific chromosome or a region on a chromosome,
and it might be useful to work with data for only that subset. This operation
is easily performed on a comparison dataframe:

```{r Subset chromosome or region}
hct116_rko_region <- hct116_rko[hct116_rko$chr == 12 &
                                hct116_rko$pos >= 25000000 &
                                hct116_rko$pos <= 30000000, ]
head(hct116_rko_region)
```

You might also be interested in a specific gene or transcript, of special
importance to your study:

```{r Subset gene or transcript}
hct116_rko_eps8_t <- hct116_rko[hct116_rko$ENSTID == "ENST00000281172", ]
hct116_rko_vamp1 <- hct116_rko[hct116_rko$ENSGID == "ENSG00000139190", ]
hct116_rko_ldhb <- hct116_rko[hct116_rko$gene == "LDHB", ]
head(hct116_rko_ldhb)
```

Here we see two mutations in the LDHB gene, one mismatching MODIFIER variant
and one matching LOW variant. This is a good way to check for known mutations
in your dataset. For example, the HCT116 cell line is supposed to have a KRAS
G13D mutation. We might look for this using its known `rsID` or position:

```{r Subset KRAS}
hct116_rko_kras <- hct116_rko[hct116_rko$rsID == "rs112445441", ]
hct116_rko_kras <- hct116_rko[hct116_rko$chr == 12 &
                              hct116_rko$pos == 25398281, ]
nrow(hct116_rko_kras)
```

This yields only empty dataframes. Why is this? The likely reason here is that
this is not a mutation present in the RKO cell line, or that there simply
wasn't enough data to get a confident variant call for this position. As the
`compare_profiles` function only looks at SNVs overlapping between two variant
profiles, variants are only seen if it is present in both profiles (or,
alternatively, a confident REF/REF call). Luckily, there is a way to check this
by simply looking for the specific variant in the originating SNV profiles
individually:

```{r Subset KRAS individually}
hct116[seqnames(hct116) == 12 & start(hct116) == 25398281, ]
rko[seqnames(rko) == 12 & start(rko) == 25398281, ]
```

Here we see that the KRAS G13D mutation is indeed present in the HCT116
profile, but not in RKO. (Please note that this data was aligned and analysed
using the GRCh37 / hg19 assembly and that listed positions might not be
accurate for subsequent assemblies.)

# Evaluating multiple comparisons

While binary analyses between two samples can tell you a lot about their
similarities, many scientific studies have more than just two, not even
mentioning meta-studies and large-scale comparisons. It is therefore important
to be able to characterise and evaluate not just binary comparisons, but also
many-to-one or many-to-many cases - the `seqCAT` package provides a
number of functions and procedures for doing so. 

## Performing multiple profile comparisons

The first step of any such analyses is to create and read SNV profiles for each
sample that is to be evaluated (please see the sections above). The example
data used here has three different samples: HCT116, HKE3 and RKO. The
`compare_many` function is a helper function for creating either one-to-many or
many-to-many SNV profile comparisons, and returns a `list` of the global
similarities for all combinations of profiles and their respective data (for
downstream analyses):

```{r Many-to-many comparisons}
# Create list of SNV profiles
profiles <- list(hct116, hke3, rko)

# Perform many-to-many comparisons
many <- compare_many(profiles)
many[[1]]
```

We can here see the summary statistics of all three combinations of the cell
lines in the example data. Notice that `compare_many` will only perform a
comparison that has not already been performed, *i.e.* it will not perform the
RKO vs. HCT116 comparison if it has already performed HCT116 vs. RKO.
Also notice that it does perform self-comparisons (*i.e.* HCT116 vs.
HCT116), which is useful for downstream visualisations.

The similarities are stored in the first element of the results (`many[[1]]`),
while the data for each comparison is stored in the second (`many[[2]]`). The
second element is itself also a list, whose indices correspond to the row names
of the similarity object. If we, for example, are interested in the HKE3
self-comparison, we can see that its row name is `4`. We can then access its
data like this:

```{r HKE3 self-comparisons}
hke3_hke3 <- many[[2]][[4]]
head(hke3_hke3)
```

You may also specify the `a` and `b` similarity score parameters, as above. If
you are interested in only a one-to-many comparison (for cases when you have a
"true" baseline profile to compare against), you can do this by also specifying
the `one = <profile>` parameter in the function call. This is useful if you
have a COSMIC profile to compare against, for example:

```{r COSMIC-to-many comparisons}
many_cosmic <- compare_many(profiles, one = cosmic)
many_cosmic[[1]]
```

It is important to note that performing many comparisons like this may take
quite some time, depending on the number of profiles and how much data each
profile has. By returning all the data in a list you may then save each
comparison to a file, for later re-analysis without having to re-do the
comparisons.

## Visualising multiple comparisons

A useful and straightforward way of visualising multiple profile comparisons is
to use a heatmap. We can use the summary statistics listed in the similarity
object from above as input to the function `plot_heatmap`, which gives you a
simple overview of all your comparisons:

```{r Plot heatmap, out.width = "60 %"}
heatmap <- plot_heatmap(many[[1]])
heatmap
```

Here we see a blue colour gradient for the similarity score of the three cell
lines, which are clustered according to their similarity (using `cluster =
TRUE`, as default). You may change the size of the text annotations using
`annotation_size = 5` (default) or suppress them entirely (`annotate = FALSE`).
You may also suppress the legend (`legend = FALSE`), change the main colour of
the gradient (`colour = "#1954A6"` by default) or change the limits of the
gradient (`limits = c(0, 50, 90, 100)` by default). The choice of gradient
limits are based on clarity (comparisons with a similarity score less than 50,
*i.e.* those that likely have too few overlapping variants to begin with, are
suppressed) and the previously mentioned 90 % concordance threshold [@Yu2015].

This heatmap makes it clear that HCT116 and HKE3 are, indeed, very similar to
each other, while RKO differs from them both. These types of heatmaps can be
created for an arbitrary number of samples, which will then give a great
overview of the global similarities of all the samples studied. This can be
used to evaluate the quality of the datasets (*e.g.* to see which comparisons
have very few overlaps), find similarity clusters and potential unexpected
outliers. If a sample stands out in a heatmap such as this, that is grounds for
further investigation, using both the methods described above and more
classical evaluations of sequencing data (read quality, adapter contamination,
alignments, variant calling, and so on).

```{r Remove temporary files, echo = FALSE, results = "hide"}
file.remove("hct116_profile.txt")
file.remove("rko_profile.txt")
file.remove("hke3_profile.txt")
```

# Citation {-}

If you are using seqCAT to analyse your samples, please cite the
article from which the methodology was first published.

Fasterius, E., Raso, C., Kennedy, S., Rauch, N., Lundin, P., Kolch, W., et al.
(2017). A novel RNA sequencing data analysis method for cell line
authentication. PloS One, 12(2), e0171435.
http://doi.org/10.1371/journal.pone.0171435

# Session info {-}

```{r Session info, echo = FALSE}
sessionInfo()
```

# References

