#' Reads variant data in 'extracted' format.
#'
#' This is a function for reading variant data in the 'extracted' type format,
#' generated by the \code{extract_variants.py} script. The read data is
#' returned as a GenomicRanges object, suitable for merging of metadata.
#' 
#' @param file The file path to be read.
#' @param sample_name The sample name from which the file originates.
#' @return A GenomicRanges object.
#' @examples
#' read_variants('~/project/data/variants/extract.txt', 'HeLa')

#' @export
read_variants = function(file, sample_name) {

    # Output message
    message(paste0('reading sample data "', basename(file), '" ...'))

    # Read data
    data = read.table(file, sep='\t', quote='\"', comment='', 
                      stringsAsFactors=FALSE)

    # Change column names
    # TODO: fix the 'extract_variants.py' script to just give correct names
    colnames = c('chr', 'pos', 'rsID', 'REF', 'ALT', 'gene', 'ENSTID', 'ENSGID',
                 'impact', 'effect', 'feature.type', 'transcript.biotype', 
                 'allele_1', 'allele_2','DP', 'filter', 'AD', 'nucleotide',
                 'amino.acid', 'warnings')
    names(data) = colnames
    data$ALT = gsub('\\[|\\]', '', data$ALT)

    # Remove duplicate variants
    data = data[!duplicated(data[, c('chr', 'pos', 'ENSGID')]), ]

    # Add sample name
    data$sample = sample_name

    # Remove indels
    data = data[(nchar(data$allele_1) == 1 & nchar(data$allele_2) == 1), ]

    # Convert to GRanges object
    data.gr = makeGRangesFromDataFrame(data, 
                                       keep.extra.columns=TRUE, 
                                       ignore.strand=TRUE,
                                       seqinfo=NULL, 
                                       seqnames.field='chr',
                                       start.field="pos",
                                       end.field="pos",
                                       starts.in.df.are.0based=FALSE)

    # Rename and remove seqlevels
    seqlevels(data.gr) = gsub("chr", "", seqlevels(data.gr))
    data.gr = 
        suppressWarnings(keepSeqlevels(data.gr, 
                                       c(as.character(1:22), 'X', 'Y')))

    # Return the GenomicRanges object
    return(data.gr)
}
