#' Reads variant data in 'extracted' format.
#'
#' This is a function for reading variant data in the 'extracted' type format,
#' generated by the \code{extract_variants.py} script. The read data is
#' returned as a GenomicRanges object, suitable for merging of metadata.
#' 
#' @param file The file path to be read.
#' @param sample_name The sample name from which the file originates.
#' @return A GenomicRanges object.
#' @examples
#' read_variants('~/project/data/variants/extract.txt', 'HeLa')

#' @export
read_variants = function(file, sample_name) {

    # Read data
    data = read.table(file, sep='\t', header=TRUE, stringsAsFactors=FALSE)

    # Remove duplicate variants
    data = data[!duplicated(data[, c('chr', 'pos', 'ENSGID')]), ]

    # Add sample name
    data$sample = sample_name

    # Convert to GRanges object
    data.gr = GenomicRanges::makeGRangesFromDataFrame(
        data, 
        keep.extra.columns=TRUE, 
        ignore.strand=TRUE,
        seqinfo=NULL, 
        seqnames.field='chr',
        start.field="pos",
        end.field="pos",
        starts.in.df.are.0based=FALSE)

    # Rename and remove seqlevels
    GenomeInfoDb::seqlevels(data.gr) = 
        gsub("chr", "", GenomeInfoDb::seqlevels(data.gr))
    data.gr = suppressWarnings(GenomeInfoDb::keepSeqlevels(
        data.gr, c(as.character(1:22), 'X', 'Y')))

    # Return the GenomicRanges object
    return(data.gr)
}
